File: src/preset.js
/**
 * Preset Module
 * Provides dynamic filter suggestions for the preset feature.
 * Each suggestion is computed from current filter option data and can be
 * applied in one click to progressively narrow down log results.
 *
 * Responsibility: Generate contextual filter suggestions based on data analytics
 */

const fs = require('fs');
const path = require('path');

/**
 * Preset Service
 * Generates dynamic filter suggestions based on current filter option data.
 * Also handles persistence of preset snapshots to/from disk.
 */
class PresetService {
  /**
   * Get path to preset snapshot JSON file
   * @param {string} dbDir - Database directory path
   * @returns {string}
   */
  static getPresetsPath(dbDir) {
    return path.join(dbDir, 'neuf-presets.json');
  }

  /**
   * Get path to user-defined preset JSON file (preset.json at project root).
   * @returns {string}
   */
  static getUserPresetsPath() {
    return path.join(__dirname, '..', 'preset.json');
  }

  /**
   * Load and parse a preset JSON file from disk.
   * Returns null if the file is absent, undefined, or cannot be parsed.
   * @param {string} filePath - Full path to the JSON file
   * @param {Function} logger - Logger function
   * @returns {Object|null} Parsed JSON object, or null on failure
   */
  static loadPresetFile(filePath, logger = console.log) {
    if (!fs.existsSync(filePath)) return null;
    try {
      const raw = fs.readFileSync(filePath, 'utf-8');
      return JSON.parse(raw) || null;
    } catch (e) {
      logger(`⚠️  Failed to load preset file "${filePath}": ${e.message}`);
      return null;
    }
  }

  /**
   * Load user-defined presets from the preset.json file at the project root.
   * Returns an empty object if the file is absent or cannot be parsed.
   * @param {string|null} jsonPath - Override path (defaults to getUserPresetsPath())
   * @param {Function} logger - Logger function
   * @returns {Object} Map of preset id to preset object
   */
  static loadUserPresets(jsonPath = null, logger = console.log) {
    const filePath = jsonPath || PresetService.getUserPresetsPath();
    const parsed = PresetService.loadPresetFile(filePath, logger);
    if (!parsed) return {};
    // Validate each entry: warn and skip any that lack a valid `filters` object,
    // because applyPreset() calls Object.entries(preset.filters) and would crash.
    const valid = {};
    for (const [id, entry] of Object.entries(parsed)) {
      if (!entry || !entry.filters || typeof entry.filters !== 'object') {
        logger(`⚠️  Skipping malformed preset "${id}" in ${filePath}: missing or invalid "filters" property`);
        continue;
      }
      valid[id] = entry;
    }
    return valid;
  }

  /**
   * Save filter options snapshot to JSON file.
   * Called once after scan so presets always reflect original data.
   * @param {string} presetsPath - Full path to the JSON file
   * @param {Object} filterOptions - Filter options from getFilterOptions()
   * @param {Function} logger - Logger function
   */
  static savePreset(presetsPath, presets, logger = console.log) {
    fs.writeFileSync(presetsPath, JSON.stringify(presets, null, 2));
    logger(`💾 Preset snapshot saved to ${presetsPath}`);
  }

  /**
   * Load filter options snapshot from JSON file and merge with user-defined presets
   * from preset.json at the project root.
   * User presets are loaded first; snapshot presets (data-derived) take precedence
   * on ID collision so dynamic suggestions always win.
   * @param {string} dbDir - path to the DB directory (used to construct full path to presets file)
   * @param {Function} logger - Logger function
   * @returns {Object} Merged preset map (never null — at minimum returns user presets or {})
   */
  static loadPreset(dbDir, logger = console.log) {
    const userPresets = PresetService.loadUserPresets(null, logger);

    const presetsPath = PresetService.getPresetsPath(dbDir);
    const snapshotPresets = PresetService.loadPresetFile(presetsPath, logger);
    if (!snapshotPresets) return userPresets;

    // Warn when a snapshot entry shadows a user-defined preset with the same id.
    // This typically means the preset was part of the codebase at scan time and
    // was written into neuf-presets.json.  Edits to preset.json won't take effect
    // for those ids until the database is rebuilt (delete neuf-logs.db and re-scan).
    const shadowed = Object.keys(userPresets).filter(id => id in snapshotPresets);
    if (shadowed.length > 0) {
      shadowed.forEach(id => logger(`⚠️  User preset "${id}" is overridden by scan preset`));
    }

    // Snapshot (data-derived) presets override user presets on key collision
    return { ...userPresets, ...snapshotPresets };
  }

  // ...existing code...

  /**
   * Merge time ranges (timeFrom/timeTo) using intersection logic.
   * Takes the larger timeFrom (most recent start) and smaller timeTo (earliest end).
   * Handles both string and numeric (Unix seconds) timestamps.
   * No-op if only one side has a time value — uses the available one.
   * @param {string|number|null} userTime - User-provided time value
   * @param {string|number|null} presetTime - Preset time value
   * @param {'from'|'to'} type - 'from' for timeFrom, 'to' for timeTo
   * @returns {string|number|null} Merged time value
   */
  static _mergeTimeRange(userTime, presetTime, type) {
    if (!userTime && !presetTime) return null;
    if (!userTime) return presetTime;
    if (!presetTime) return userTime;

    // Convert to comparable format (Unix seconds)
    const userSec = typeof userTime === 'number' ? userTime : this._parseTimestampSec(userTime)?.floorSec;
    const presetSec = typeof presetTime === 'number' ? presetTime : this._parseTimestampSec(presetTime)?.floorSec;

    if (userSec === null || userSec === undefined || presetSec === null || presetSec === undefined) {
      // If parse fails, return user value (user input takes precedence in case of error)
      return userTime;
    }

    // For timeFrom: take the larger (more recent start) — creates narrower range
    // For timeTo: take the smaller (earlier end) — creates narrower range
    const selectedSec = type === 'from' ? Math.max(userSec, presetSec) : Math.min(userSec, presetSec);

    // Return in same format as userTime (preserve original format)
    if (typeof userTime === 'number') {
      return selectedSec;
    } else {
      return this._formatUnixSec(selectedSec);
    }
  }

  /**
   * Apply preset filters into an existing filters object (mutates in place).
   * Reads filters.preset (string or array), resolves against snapshotOptions,
   * and merges matching preset filters in order.
   * For time ranges (timeFrom/timeTo): computes intersection when both preset and user have values.
   * For array fields: merges values (combines multiple presets).
   * For other fields: uses preset value if not already set by user.
   * No-op if filters.preset is empty or no preset matches.
   * @param {Object} filters - Filters object (will be mutated in place)
   * @param {Object} presets - filterOptions presets from base logs table
   * @param {Function} logger - Logger function
   */
  static applyPreset(filters, presets, logger = console.log) {
    if (!filters || !filters.preset) return;

    const presetIds = Array.isArray(filters.preset)
      ? filters.preset.filter(Boolean)
      : [filters.preset];

    if (presetIds.length === 0) return;

    for (const presetId of presetIds) {
      const suggestion = presets[presetId];
      if (!suggestion) {
        logger(`⚠️  Preset not found: "${presetId}"`);
        continue;
      }
      // Merge preset filters into user filters
      for (const [key, value] of Object.entries(suggestion.filters)) {
        if (key === 'timeFrom' || key === 'timeTo') {
          // Time range: compute intersection
          filters[key] = this._mergeTimeRange(filters[key], value, key === 'timeFrom' ? 'from' : 'to');
        } else if (Array.isArray(value) && Array.isArray(filters[key])) {
          // Array field: merge values to support combining multiple presets
          filters[key] = [...new Set([...filters[key], ...value])];
        } else {
          // Other fields: use preset value if not already set
          if (filters[key] === undefined || filters[key] === null || filters[key] === '') {
            filters[key] = value;
          }
        }
      }
    }
  }

  /**
   * Parse a log timestamp string to floor/ceil Unix seconds.
   * Supports formats with and without milliseconds:
   * - "YYYY.MM.DD HH:mm:ss.SSS" (with milliseconds)
   * - "YYYY.MM.DD HH:mm:ss" (without milliseconds)
   * @param {string} ts - Timestamp string
   * @returns {{ floorSec: number, ceilSec: number }|null}
   */
  static _parseTimestampSec(ts) {
    // Match with optional milliseconds
    const m = ts.match(/^(\d{4})\.(\d{2})\.(\d{2}) (\d{2}):(\d{2}):(\d{2})(?:\.(\d{3}))?/);
    if (!m) return null;
    const ms = m[7] ? parseInt(m[7], 10) : 0;
    const unixMs = Date.UTC(
      parseInt(m[1], 10), parseInt(m[2], 10) - 1, parseInt(m[3], 10),
      parseInt(m[4], 10), parseInt(m[5], 10), parseInt(m[6], 10), ms
    );
    return {
      floorSec: Math.floor(unixMs / 1000),
      ceilSec: Math.ceil(unixMs / 1000)
    };
  }

  /**
   * Format Unix timestamp (seconds, UTC) to "YYYY.MM.DD HH:mm:ss" string.
   * @param {number} unixSec
   * @returns {string}
   */
  static _formatUnixSec(unixSec) {
    const d = new Date(unixSec * 1000);
    const pad = n => String(n).padStart(2, '0');
    return `${d.getUTCFullYear()}.${pad(d.getUTCMonth() + 1)}.${pad(d.getUTCDate())} ${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}`;
  }

  /**
   * Create install-based presets from LifecycleProvider install log entries.
   * Groups by device_id, sorts timestamps, and creates a preset for each interval
   * between consecutive installs plus one for the last (open-ended) install.
   * timeFrom/timeTo stored as Unix seconds (numbers) so normalizeFilters passes them through unchanged.
   * @param {Array<{device_id: string, timestamp: string}>} installLogs - Install log rows from DB
   * @returns {Object} Map of preset id to preset object
   */
  static createInstallPresets(installLogs) {
    const presets = {};

    // Group timestamps by device_id
    const byDevice = {};
    for (const log of installLogs) {
      if (!log.device_id) continue;
      if (!byDevice[log.device_id]) byDevice[log.device_id] = [];
      byDevice[log.device_id].push(log.timestamp);
    }

    for (const [deviceId, timestamps] of Object.entries(byDevice)) {
      // Ensure ascending order
      timestamps.sort();

      for (let i = 0; i < timestamps.length; i++) {
        const parsedFrom = PresetService._parseTimestampSec(timestamps[i]);
        if (!parsedFrom) continue;

        const timeFromSec = parsedFrom.floorSec;
        const timeFromStr = PresetService._formatUnixSec(timeFromSec);
        const presetId = `fujifilm_${deviceId}_install_${i + 1}`;

        let label, filters;

        if (i < timestamps.length - 1) {
          // Interval between this install and the next
          const parsedTo = PresetService._parseTimestampSec(timestamps[i + 1]);
          if (!parsedTo) continue;
          const timeToSec = parsedTo.ceilSec;
          const timeToStr = PresetService._formatUnixSec(timeToSec);
          label = `📦 Fujifilm-${deviceId} install from ${timeFromStr} -> ${timeToStr}`;
          filters = { timeFrom: timeFromSec, timeTo: timeToSec, deviceInclude: [deviceId, null] };
        } else {
          // Last install — open-ended, no timeTo
          label = `📦 Fujifilm-${deviceId} install from ${timeFromStr}`;
          filters = { timeFrom: timeFromSec, deviceInclude: [deviceId, null] };
        }

        presets[presetId] = {
          id: presetId,
          label,
          description: label,
          filters
        };
      }
    }

    return presets;
  }
}

module.exports = { PresetService };

File: python-port/src/preset.py
"""
Preset Module (Python port of src/preset.js)
Provides dynamic filter suggestions for the preset feature.
Responsibility: Generate contextual filter suggestions based on data analytics.
"""

import json
import os
import re
import calendar


class PresetService:
    """
    Preset Service.
    Generates dynamic filter suggestions and manages preset persistence.
    All methods are static, mirroring the JavaScript PresetService.
    """

    @staticmethod
    def get_presets_path(filter_db_dir):
        """Return full path to the scan-time preset snapshot JSON file."""
        return os.path.join(filter_db_dir, 'neuf-presets.json')

    @staticmethod
    def get_user_presets_path():
        """Return full path to the user-defined preset.json at the project root."""
        # python-port/src/ → up two levels → project root
        return os.path.join(
            os.path.dirname(__file__), '..', '..', 'preset.json'
        )

    @staticmethod
    def load_preset_file(file_path, logger=print):
        """Load and parse a single preset JSON file from disk."""
        if not os.path.exists(file_path):
            return None
        try:
            with open(file_path, encoding='utf-8') as fh:
                return json.load(fh) or None
        except Exception as e:
            logger(f'⚠️  Failed to load preset file "{file_path}": {e}')
            return None

    @staticmethod
    def load_user_presets(json_path=None, logger=print):
        """
        Load user-defined presets from preset.json.
        Returns empty dict if absent or unparseable.
        Validates each entry has a 'filters' dict; warns and skips malformed entries.
        """
        file_path = json_path if json_path is not None \
                    else PresetService.get_user_presets_path()
        parsed = PresetService.load_preset_file(file_path, logger)
        if not parsed:
            return {}

        valid = {}
        for preset_id, entry in parsed.items():
            if (not entry or
                    not entry.get('filters') or
                    not isinstance(entry.get('filters'), dict)):
                logger(
                    f'⚠️  Skipping malformed preset "{preset_id}" in {file_path}: '
                    f'missing or invalid "filters" property'
                )
                continue
            valid[preset_id] = entry
        return valid

    @staticmethod
    def save_preset(presets_path, presets, logger=print):
        """Persist a presets dict to disk as JSON."""
        with open(presets_path, 'w', encoding='utf-8') as fh:
            json.dump(presets, fh, indent=2)
        logger(f'💾 Preset snapshot saved to {presets_path}')

    @staticmethod
    def load_preset(filter_db_dir, logger=print):
        """
        Load and merge user presets (preset.json) with scan-time snapshot presets.
        Snapshot presets take precedence on id collision.
        """
        user_presets = PresetService.load_user_presets(None, logger)

        presets_path = PresetService.get_presets_path(filter_db_dir)
        snapshot = PresetService.load_preset_file(presets_path, logger)
        if not snapshot:
            return user_presets

        # Warn when snapshot shadows user preset
        shadowed = [pid for pid in user_presets if pid in snapshot]
        for pid in shadowed:
            logger(f'⚠️  User preset "{pid}" is overridden by scan preset')

        return {**user_presets, **snapshot}

    # ------------------------------------------------------------------ #
    #  Time range helpers                                                   #
    # ------------------------------------------------------------------ #

    @staticmethod
    def _merge_time_range(user_time, preset_time, range_type):
        """
        Intersect a single timeFrom or timeTo value from user + preset.
        """
        if not user_time and not preset_time:
            return None
        if not user_time:
            return preset_time
        if not preset_time:
            return user_time

        user_sec = (user_time if isinstance(user_time, (int, float))
                    else (PresetService._parse_timestamp_sec(user_time) or {}).get('floorSec'))
        preset_sec = (preset_time if isinstance(preset_time, (int, float))
                      else (PresetService._parse_timestamp_sec(preset_time) or {}).get('floorSec'))

        if user_sec is None or preset_sec is None:
            return user_time

        selected_sec = max(user_sec, preset_sec) if range_type == 'from' \
                       else min(user_sec, preset_sec)

        if isinstance(user_time, (int, float)):
            return selected_sec
        return PresetService._format_unix_sec(selected_sec)

    @staticmethod
    def apply_preset(filters, presets, logger=print):
        """
        Mutate filters dict by merging filters from one or more preset IDs.
        """
        if not filters or 'preset' not in filters:
            return

        raw = filters['preset']
        preset_ids = [p for p in (raw if isinstance(raw, list) else [raw]) if p]

        if not preset_ids:
            return

        for preset_id in preset_ids:
            suggestion = presets.get(preset_id)
            if not suggestion:
                logger(f'⚠️  Preset not found: "{preset_id}"')
                continue

            for key, value in suggestion['filters'].items():
                if key in ('timeFrom', 'timeTo'):
                    filters[key] = PresetService._merge_time_range(
                        filters.get(key), value,
                        'from' if key == 'timeFrom' else 'to'
                    )
                elif isinstance(value, list) and isinstance(filters.get(key), list):
                    # Merge without duplicates, preserving order
                    seen = []
                    for v in [*filters[key], *value]:
                        if v not in seen:
                            seen.append(v)
                    filters[key] = seen
                else:
                    if filters.get(key) is None or filters.get(key) == '' or key not in filters:
                        filters[key] = value

    # ------------------------------------------------------------------ #
    #  Timestamp helpers                                                    #
    # ------------------------------------------------------------------ #

    _TS_RE = re.compile(
        r'^(\d{4})\.(\d{2})\.(\d{2}) (\d{2}):(\d{2}):(\d{2})(?:\.(\d{3}))?'
    )

    @staticmethod
    def _parse_timestamp_sec(ts):
        """Parse NEUF timestamp string to floor/ceil Unix seconds."""
        m = PresetService._TS_RE.match(ts)
        if not m:
            return None
        ms = int(m.group(7)) if m.group(7) else 0
        unix_ms = int(calendar.timegm((
            int(m.group(1)), int(m.group(2)), int(m.group(3)),
            int(m.group(4)), int(m.group(5)), int(m.group(6)),
            0, 0, 0
        ))) * 1000 + ms
        return {
            'floorSec': unix_ms // 1000,
            'ceilSec':  (unix_ms + 999) // 1000,
        }

    @staticmethod
    def _format_unix_sec(unix_sec):
        """Format Unix seconds (UTC) to "YYYY.MM.DD HH:mm:ss" string."""
        import datetime
        d = datetime.datetime.utcfromtimestamp(unix_sec)
        return (f'{d.year:04d}.{d.month:02d}.{d.day:02d}'
                f' {d.hour:02d}:{d.minute:02d}:{d.second:02d}')

    @staticmethod
    def create_install_presets(install_logs):
        """
        Generate install-based presets from LifecycleProvider install log rows.
        """
        presets = {}

        # Group timestamps by device_id
        by_device = {}
        for log in install_logs:
            dev = log.get('device_id')
            if not dev:
                continue
            by_device.setdefault(dev, []).append(log['timestamp'])

        for device_id, timestamps in by_device.items():
            timestamps.sort()

            for i, ts in enumerate(timestamps):
                parsed_from = PresetService._parse_timestamp_sec(ts)
                if not parsed_from:
                    continue

                time_from_sec = parsed_from['floorSec']
                time_from_str = PresetService._format_unix_sec(time_from_sec)
                preset_id = f'fujifilm_{device_id}_install_{i + 1}'

                if i < len(timestamps) - 1:
                    parsed_to = PresetService._parse_timestamp_sec(timestamps[i + 1])
                    if not parsed_to:
                        continue
                    time_to_sec = parsed_to['ceilSec']
                    time_to_str = PresetService._format_unix_sec(time_to_sec)
                    label = (f'📦 Fujifilm-{device_id} install from '
                             f'{time_from_str} -> {time_to_str}')
                    filt = {
                        'timeFrom': time_from_sec,
                        'timeTo': time_to_sec,
                        'deviceInclude': [device_id, None],
                    }
                else:
                    label = f'📦 Fujifilm-{device_id} install from {time_from_str}'
                    filt = {'timeFrom': time_from_sec, 'deviceInclude': [device_id, None]}

                presets[preset_id] = {
                    'id': preset_id,
                    'label': label,
                    'description': label,
                    'filters': filt,
                }

        return presets
